quinn-udp/src/unix.rs

```diff
@@ -442,25 +442,9 @@ unsafe fn recvmmsg_with_fallback(
     let flags = 0;
     let timeout = ptr::null_mut::<libc::timespec>();
 
-    #[cfg(not(any(target_os = "freebsd", target_os = "netbsd")))]
-    {
-        let ret =
-            libc::syscall(libc::SYS_recvmmsg, sockfd, msgvec, vlen, flags, timeout) as libc::c_int;
-        if ret != -1 {
-            return ret;
-        }
-    }
-
-    // libc on FreeBSD and NetBSD implement `recvmmsg` as a high-level abstraction over
-    // `recvmsg`, thus `SYS_recvmmsg` constant and direct system call do not exist
-    #[cfg(any(target_os = "freebsd", target_os = "netbsd"))]
-    {
-        #[cfg(target_os = "freebsd")]
-        let vlen = vlen as usize;
-        let ret = libc::recvmmsg(sockfd, msgvec, vlen, flags, timeout) as libc::c_int;
-        if ret != -1 {
-            return ret;
-        }
+    let ret = libc::recvmmsg(sockfd, msgvec, vlen as _, flags, timeout) as libc::c_int;
```

Comment:
Note the `as _` here. The alternative would be:

``` diff
modified   quinn-udp/src/unix.rs
@@ -442,7 +442,11 @@ unsafe fn recvmmsg_with_fallback(
     let flags = 0;
     let timeout = ptr::null_mut::<libc::timespec>();
 
-    let ret = libc::recvmmsg(sockfd, msgvec, vlen as _, flags, timeout) as libc::c_int;
+    let ret = {
+        #[cfg(target_os = "freebsd")]
+        let vlen = vlen as usize;
+        libc::recvmmsg(sockfd, msgvec, vlen, flags, timeout) as libc::c_int
+    };
     if ret != -1 {
         return ret;
     }
```

The additional scope is need in order to preserve the original `vlen` type (`c_int`) (i.e. not shadow `vlen`) for the `recvmmsg_fallback` below.

Reply:
Or we don't pass `vlen` to `recvmmsg_fallback` at all:

``` diff
modified   quinn-udp/src/unix.rs
@@ -181,6 +181,10 @@ impl UdpSocketState {
         bufs: &mut [IoSliceMut<'_>],
         meta: &mut [RecvMeta],
     ) -> io::Result<usize> {
+        if bufs.len() == 0 {
+            return Ok(0);
+        }
+
         recv(socket.0, bufs, meta)
     }
 
@@ -442,7 +446,9 @@ unsafe fn recvmmsg_with_fallback(
     let flags = 0;
     let timeout = ptr::null_mut::<libc::timespec>();
 
-    let ret = libc::recvmmsg(sockfd, msgvec, vlen as _, flags, timeout) as libc::c_int;
+    #[cfg(target_os = "freebsd")]
+    let vlen = vlen as usize;
+    let ret = libc::recvmmsg(sockfd, msgvec, vlen, flags, timeout) as libc::c_int;
     if ret != -1 {
         return ret;
     }
@@ -451,7 +457,7 @@ unsafe fn recvmmsg_with_fallback(
     match e.raw_os_error() {
         Some(libc::ENOSYS) => {
             // Fallback to `recvmsg`.
-            recvmmsg_fallback(sockfd, msgvec, vlen)
+            recvmmsg_fallback(sockfd, msgvec)
         }
         _ => -1,
     }
@@ -461,15 +467,8 @@ unsafe fn recvmmsg_with_fallback(
 /// for systems which do not support `recvmmsg`
 /// such as Linux <2.6.33.
 #[cfg(not(any(target_os = "macos", target_os = "ios", target_os = "openbsd")))]
-unsafe fn recvmmsg_fallback(
-    sockfd: libc::c_int,
-    msgvec: *mut libc::mmsghdr,
-    vlen: libc::c_uint,
-) -> libc::c_int {
+unsafe fn recvmmsg_fallback(sockfd: libc::c_int, msgvec: *mut libc::mmsghdr) -> libc::c_int {
     let flags = 0;
-    if vlen == 0 {
-        return 0;
-    }
 
     let n = libc::recvmsg(sockfd, &mut (*msgvec).msg_hdr, flags);
     if n == -1 {
```

Your call. I am fine either way.